File: examples/03-composition/python/inventory-service/main.py
import logging
import os
from concurrent import futures

import grpc
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.grpc import GrpcInstrumentorServer
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

import inventory_pb2
import inventory_pb2_grpc

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("inventory")

OTEL_ENDPOINT = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://lgtm:4318")

resource = Resource.create({"service.name": "inventory"})
provider = TracerProvider(resource=resource)
provider.add_span_processor(
    BatchSpanProcessor(OTLPSpanExporter(endpoint=f"{OTEL_ENDPOINT}/v1/traces"))
)
trace.set_tracer_provider(provider)
GrpcInstrumentorServer().instrument()

STOCK = {
    "widget-a": 42,
    "widget-b": 17,
    "gadget-x": 100,
    "gadget-y": 0,
}
call_count = 0


class InventoryServicer(inventory_pb2_grpc.InventoryServicer):
    def GetStock(self, request, context):
        global call_count
        call_count += 1
        available = STOCK.get(request.sku, 0)
        log.info("GetStock sku=%s available=%d (call #%d)", request.sku, available, call_count)
        return inventory_pb2.GetStockReply(sku=request.sku, available=available)

    def GetStockBatch(self, request, context):
        global call_count
        call_count += 1
        items = []
        for sku in request.skus:
            available = STOCK.get(sku, 0)
            items.append(inventory_pb2.GetStockReply(sku=sku, available=available))
        log.info("GetStockBatch skus=%s (call #%d)", list(request.skus), call_count)
        return inventory_pb2.GetStockBatchReply(items=items)


def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=4))
    inventory_pb2_grpc.add_InventoryServicer_to_server(InventoryServicer(), server)
    server.add_insecure_port("[::]:50051")
    server.start()
    log.info("inventory gRPC server listening on :50051")
    server.wait_for_termination()


if __name__ == "__main__":
    serve()

File: examples/03-composition/cpp/inventory-service/src/main.cpp
#include <grpcpp/grpcpp.h>
#include "inventory.grpc.pb.h"
#include <unordered_map>
#include <iostream>
#include <atomic>

static std::unordered_map<std::string, int> g_stock = {
    {"widget-a", 42},
    {"widget-b", 17},
    {"gadget-x", 100},
    {"gadget-y", 0},
};
static std::atomic<int> g_call_count{0};

class InventoryServiceImpl final : public inventory::Inventory::Service {
    grpc::Status GetStock(grpc::ServerContext*,
                          const inventory::GetStockRequest* req,
                          inventory::GetStockReply* reply) override {
        int count = ++g_call_count;
        auto it = g_stock.find(req->sku());
        int available = (it != g_stock.end()) ? it->second : 0;
        reply->set_sku(req->sku());
        reply->set_available(available);
        std::cerr << "GetStock sku=" << req->sku()
                  << " available=" << available
                  << " (call #" << count << ")" << std::endl;
        return grpc::Status::OK;
    }

    grpc::Status GetStockBatch(grpc::ServerContext*,
                               const inventory::GetStockBatchRequest* req,
                               inventory::GetStockBatchReply* reply) override {
        int count = ++g_call_count;
        for (const auto& sku : req->skus()) {
            auto* item = reply->add_items();
            item->set_sku(sku);
            auto it = g_stock.find(sku);
            item->set_available((it != g_stock.end()) ? it->second : 0);
        }
        std::cerr << "GetStockBatch skus=[";
        for (int i = 0; i < req->skus_size(); i++) {
            if (i > 0) std::cerr << ", ";
            std::cerr << req->skus(i);
        }
        std::cerr << "] (call #" << count << ")" << std::endl;
        return grpc::Status::OK;
    }
};

int main() {
    std::string addr("0.0.0.0:50051");
    InventoryServiceImpl service;
    grpc::ServerBuilder builder;
    builder.AddListeningPort(addr, grpc::InsecureServerCredentials());
    builder.RegisterService(&service);
    auto server = builder.BuildAndStart();
    std::cerr << "inventory gRPC server listening on " << addr << std::endl;
    server->Wait();
    return 0;
}
